### DISCUSSION TITLE
Shelly Dimmer 2 - How to upload original STM32 firmware

### DISCUSSION THREAD

**Discussion by CReimer** [arendst/Tasmota#11050] [Q&A] [answered]
I'd like to upload and use the original Shelly STM32 firmware with Tasmota. Because I might be able to adapt Tasmota to the original firmware, but I'm definetly unable to fix jamesturton's firmware.

I already figured out how to extract the .bin file from the original Shelly firmware bundle, but Tasmota refuses to upload it to the stm32.

Can someone point me in the right direction here? How does Tasmota figure out, which binary is uploaded? Are there any size limitations? The original stm32 firmware is 38 kB as of version 1.9.5

Thanks

Christopher

**Comment by barbudor** [comment, marked as answer]
The Shelly firmware is not freely available

The simplest way is to 1st revert to full Shelly using Tasmota-to-mgos and restart from here

**Reply by barbudor** [reply]
If you expect to solve no-neutral, my own experience is that even Shelly's firmware is far from perfect.

**Reply by CReimer** [reply]
It's not. But it doesn't have to be.
I already have that part extracted from the original bundle. Inside the original zip file is a fs.bin. This can be extracted with unspiffs8. Inside the fs.bin is a stm.bin file. Which I believe is what needs to be flashed to the stm32.

For starters I'd like to figure out how I can make Tasmota recognize this binary file and have it flashed to the stm32.

**Reply by CReimer** [reply]
Yes. I expect to solve no-neutral with the original stm32 firmware. Which in my experience works perfectly fine with the original Shelly firmware. Nothing to complain about, except that I'd like to have Tasmota in the ESP part

**Reply by barbudor** [reply]
The upload is managed from `xdrv_01_webserver.ino` and the following test is performed to identify a Shelly Dimmer firmware:
```
#ifdef SHELLY_FW_UPGRADE
      else if (ShdPresent() && (0x00 == upload.buf[0]) && (0x10 == upload.buf[1])) {
        BUploadInit(UPL_SHD);
      }
#endif  // SHELLY_FW_UPGRADE
```

If your bin doesn't have the proper fingerprint, Tasmota will not upload.

**Reply by CReimer** [reply]
Thanks. That's exactly what I was looking for

**Reply by barbudor** [reply]
Hi @CReimer 
Did you had any success in making Tasmota to work with the original STM32 firmware ?

**Reply by CReimer** [reply]
Yes and no. It should work with neutral wire, but not with Shelly Bypass. With Shelly Bypass Dimming Range stops at around 70%. Above that Tasmota keeps rebooting.

I had to modify Tasmota quite a bit, because there are some small differences between James' STM firmware and the official one.

If you want I can clean up what I have and upload it to Github.

**Reply by barbudor** [reply]
So what did you choose finally ? You mods or revert to Shelly's firmware ?
It would be interesting to have a #define in Tasmota to let people choose between original and Jame's
A failsafe on the 70% limit could be implemented as a remap of tasmota's 0..100% range to a Shelly's 0..70%

**Reply by CReimer** [reply]
I chose the original Shelly Firmware. For some reason it can dim much closer to 100% without the reboot problem. I couldn't figure out why it happens, and how Allterco works around it

**Reply by CReimer** [reply]
Here's what I have so far. https://github.com/CReimer/Tasmota

It does work with the original firmware and it should still work with James' firmware. It does not work reliably with Shelly Bypass and I couldn't figure out why.

**Reply by arendst** [reply]
Is the change only in ``xdrv_01_webserver.ino`` and ``xdrv_45_shelly_dimmer.ino`` ?

Any chance to make the Shelly firmware available for download?

**Reply by CReimer** [reply]
Yes that is all. xdrv_01 for allowing original stm.bin file upload and xdrv_45 to make communicaiton work.

I'm not sure about the firmware being available for download. You'd have to ask Allterco for permission I think.

But the latest firmware zip is always available here: http://shelly-api-eu.shelly.cloud/firmware/SHDM-2.zip
Inside is a fs.bin file which needs to be unpacked using unspiffs8 (https://github.com/ioprev/shelly-firmware)
And inside that is a stm.bin file which is exactly the file you need.

**Reply by arendst** [reply]
Thx. I managed to unspiff it :-)

For now let's use this procedure as I expect Alterco not to allow direct downloads of the stm32 file as it could easily change over time within their fs.bin
